tokio/src/sync/watch.rs

```diff
@@ -461,8 +462,14 @@ impl<T> Sender<T> {
         {
             // Acquire the write lock and update the value.
             let mut lock = self.shared.value.write().unwrap();
-            // Update the value.
-            func(&mut lock);
+            // Update the value and catch possible panic inside func.
+            let result = panic::catch_unwind(panic::AssertUnwindSafe(|| {
+                func(&mut lock);
+            }));
+            // If the func panicked return the panic to the caller.
+            if let Err(error) = result {
+                panic::resume_unwind(error);
+            }
```

Comment:
This would still poison the lock. You have to drop it before you resume unwinding.